Contracts: Rework host fn benchmarks

pgherveou opened the pull request:
fix https://github.com/paritytech/polkadot-sdk/issues/4163

This PR does the following:
Update to pallet-contracts-proc-macro: 
- Parse #[cfg] so we can add a dummy noop host function for benchmark.
- Generate BenchEnv::<host_fn> so we can call host functions directly in the benchmark.
- Add the weight of the noop host function before calling the host function itself

Update benchmarks:
- Update all host function benchmark, a host function benchmark now simply call the host function, instead of invoking the function n times from within a contract.
- Refactor RuntimeCosts & Schedule, for most host functions, we can now use the generated weight function directly instead of computing the diff with the cost! macro

```rust
// Before
#[benchmark(pov_mode = Measured)]
fn seal_input(r: Linear<0, API_BENCHMARK_RUNS>) {
    let code = WasmModule::<T>::from(ModuleDefinition {
        memory: Some(ImportedMemory::max::<T>()),
        imported_functions: vec![ImportedFunction {
            module: "seal0",
            name: "seal_input",
            params: vec![ValueType::I32, ValueType::I32],
            return_type: None,
        }],
        data_segments: vec![DataSegment { offset: 0, value: 0u32.to_le_bytes().to_vec() }],
        call_body: Some(body::repeated(
            r,
            &[
                Instruction::I32Const(4), // ptr where to store output
                Instruction::I32Const(0), // ptr to length
                Instruction::Call(0),
            ],
        )),
        ..Default::default()
    });

    call_builder!(func, code);

    let res;
    #[block]
    {
        res = func.call();
    }
    assert_eq!(res.did_revert(), false);
}
```

```rust
// After
fn seal_input(n: Linear<0, { code::max_pages::<T>() * 64 * 1024 - 4 }>) {
    let mut setup = CallSetup::<T>::default();
    let (mut ext, _) = setup.ext();
    let mut runtime = crate::wasm::Runtime::new(&mut ext, vec![42u8; n as usize]);
    let mut memory = memory!(n.to_le_bytes(), vec![0u8; n as usize],);
    let result;
    #[block]
    {
        result = BenchEnv::seal0_input(&mut runtime, &mut memory, 4, 0)
    }
    assert_ok!(result);
    assert_eq!(&memory[4..], &vec![42u8; n as usize]);
}
``` 

[Weights compare](https://weights.tasty.limo/compare?unit=weight&ignore_errors=true&threshold=10&method=asymptotic&repo=polkadot-sdk&old=master&new=pg%2Frework-host-benchs&path_pattern=substrate%2Fframe%2Fcontracts%2Fsrc%2Fweights.rs%2Cpolkadot%2Fruntime%2F*%2Fsrc%2Fweights%2F**%2F*.rs%2Cpolkadot%2Fbridges%2Fmodules%2F*%2Fsrc%2Fweights.rs%2Ccumulus%2F**%2Fweights%2F*.rs%2Ccumulus%2F**%2Fweights%2Fxcm%2F*.rs%2Ccumulus%2F**%2Fsrc%2Fweights.rs)

pgherveou commented on substrate/frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -629,6 +629,27 @@ mod benchmarks {
+		let mut setup = CallSetup::<T>::default();
+		let (mut ext, _) = setup.ext();
+		let mut runtime = crate::wasm::Runtime::new(&mut ext, vec![]);
+		let mut memory = [0u8; 36];
+		memory[0..4].copy_from_slice(&32u32.to_le_bytes());
+
+		let result;
+		#[block]
+		{
+			result = crate::wasm::BenchEnv::seal0_caller(&mut runtime, &mut memory, 4, 0);
+		}
```
> One solution would be to make [Runtime](https://github.com/paritytech/polkadot-sdk/blob/e81322bc3e8192b536067fed3ef9e20f2752c376/substrate/frame/contracts/src/wasm/runtime.rs#L375-L380) generic over the memory access so we can mock it from the benchmarks.

@athei do we really need to make memory generic? Doesn't this approach above just work?

pgherveou replied:
ah nvm I guess we need it so that we abstract gas counting for when we read and write to memory sandbox

pgherveou commented:
bot bench substrate-pallet --pallet=pallet_contracts

pgherveou commented:
bot bench substrate-pallet --pallet=pallet_contracts

athei reviewed (commented):
At first glance it looks like weights go down. But when looking at the formular we can see that the weights go up for many host functions:
<img width="1672" alt="Screenshot 2024-05-07 at 12 50 02" src="https://github.com/paritytech/polkadot-sdk/assets/2580396/9689c9f1-5e6c-46a7-ab1f-57e098755f93">

Before it was 280k and not it is 680k. They kind of should go down because no we are running the host functions directly.

pgherveou commented:
> At first glance it looks like weights go down. But when looking at the formular we can see that the weights go up for many host functions: <img alt="Screenshot 2024-05-07 at 12 50 02" width="1672" src="https://private-user-images.githubusercontent.com/2580396/328493909-9689c9f1-5e6c-46a7-ab1f-57e098755f93.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwODIxNzAsIm5iZiI6MTcxNTA4MTg3MCwicGF0aCI6Ii8yNTgwMzk2LzMyODQ5MzkwOS05Njg5YzlmMS01ZTZjLTQ2YTctYWIxZi01N2UwOTg3NTVmOTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDdUMTEzNzUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGU3NDQyYTRhMmI0ZWVlYTYxZjkxYjBlNmRlMzViY2I1ZmZlYWE3YmY4M2NhNGI3YjVmOGVlNjVlZjUzZTQwYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.f6Bs3UbGV49zuyXGpJQvg35FvIj2bNvC8u_FZPprE_M">
> 
> https://weights.tasty.limo/compare?unit=weight&ignore_errors=true&threshold=10&method=asymptotic&repo=polkadot-sdk&old=master&new=pg%2Frework-host-benchs&path_pattern=substrate%2Fframe%2Fcontracts%2Fsrc%2Fweights.rs%2Cpolkadot%2Fruntime%2F*%2Fsrc%2Fweights%2F**%2F*.rs%2Cpolkadot%2Fbridges%2Fmodules%2F*%2Fsrc%2Fweights.rs%2Ccumulus%2F**%2Fweights%2F*.rs%2Ccumulus%2F**%2Fweights%2Fxcm%2F*.rs%2Ccumulus%2F**%2Fsrc%2Fweights.rs
> 
> Before it was 280k and not it is 680k. They kind of should go down because no we are running the host functions directly.

Not sure if it should go down, but it should be roughly equal for sure.
To be correct you would need to add as well the noop_host_fn weight which adds another 70k, since the previous measured time was (fn time + base cost of calling a host fn) repeated r times.

athei commented:
Yes it should be roughly the same. Even slightly less after the rework because we still add the `noop_host_fn` onto the new weight as you pointed out.

Could be attributed to caching behaviour. Before we ran the same code over and over again. That could make each of them faster.

pgherveou commented:
Yes exactly for host fn where there is io, there are some trie node cached that amortized the cost but for other like gas_left where there are no PoV this must be something else 🤔

athei reviewed (changes requested):
For the benchmarks which touch storage (`seal_call, `seal_instantiate`, `seal_delegate_call`, ...) we need to whitelist the callers address and probably other keys.

pgherveou commented on substrate/frame/contracts/src/wasm/runtime.rs:
```diff
@@ -256,85 +268,58 @@ impl<T: Config> Token<T> for RuntimeCosts {
+			Random => T::WeightInfo::seal_random(),
+			DepositEvent { num_topic, len } => T::WeightInfo::seal_deposit_event(num_topic, len),
+			DebugMessage(len) => T::WeightInfo::seal_debug_message(len),
+			SetStorage { new_bytes, old_bytes } =>
+				T::WeightInfo::seal_set_storage_per_new_byte(new_bytes)
+					.saturating_add(T::WeightInfo::seal_set_storage_per_old_byte(old_bytes)),
```
right good point

pgherveou commented:
> For the benchmarks which touch storage (`seal_call, `seal_instantiate`, `seal_delegate_call`, ...) we need to whitelist the callers address and probably other keys.

Added that to the host fn benchmark setup.
Storage access for the contract account_id and  ContractInfo (that is touched when we add / pop a frame) are already taken into account with the base call extrinsic benchmark.

Everything else should be fine (will double check once the benchmark are updated)

I have also removed from the kitchensink in this PR #4489 the dynamic parameters that should not be used with pallet-contracts until there is a proper way to not double count them in the host fn benchmarks

```rs
		// Whitelist contract account, as it is already accounted for in the call benchmark
		benchmarking::add_to_whitelist(
			frame_system::Account::<T>::hashed_key_for(&contract.account_id).into(),
		);

		// Whitelist the contract's contractInfo as it is already accounted for in the call
		// benchmark
		benchmarking::add_to_whitelist(
			crate::ContractInfoOf::<T>::hashed_key_for(&contract.account_id).into(),
		);
```

pgherveou commented:
bot bench substrate-pallet --pallet=pallet_contracts

pgherveou commented:
bot bench substrate-pallet --pallet=pallet_contracts

pgherveou commented:
bot cancel 1-621accc2-6fcb-4497-a21d-db259c223c49

pgherveou commented:
bot bench substrate-pallet --pallet=pallet_contracts

athei commented:
<img width="1031" alt="Screenshot 2024-05-17 at 12 50 08" src="https://github.com/paritytech/polkadot-sdk/assets/2580396/3de6ff20-2d77-4c46-93b1-a6034393ead4">

`seal_call` still has **3** extra reads and **1** extra write from before.

pgherveou commented:
> <img alt="Screenshot 2024-05-17 at 12 50 08" width="1031" src="https://private-user-images.githubusercontent.com/2580396/331566074-3de6ff20-2d77-4c46-93b1-a6034393ead4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NDM2MzQsIm5iZiI6MTcxNTk0MzMzNCwicGF0aCI6Ii8yNTgwMzk2LzMzMTU2NjA3NC0zZGU2ZmYyMC0yZDc3LTRjNDYtOTNiMS1hNjAzNDM5M2VhZDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTdUMTA1NTM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWUzMTNhM2Y2NjRkZDU2M2IzMmE5OGJiOWU3ZGE2MGIxYmVmOTI2MDRlZWE3MTM0NmExNjQzZmI1MGQ5MDkxNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.O0CyK0TPsm41OWFhgotdS0ncurv76YqYdOkeYppk8qc">
> `seal_call` still has **3** extra reads and **1** extra write from before.

I traced the execution this seems correct, let me check the old equation

athei reviewed (commented):
Yes you are right. The storage accesses look correct. They were too low before.  One last thing. There should be only one access to `Account` for `seal_instantiate` to bring it in line with `seal_call`.

<img width="1369" alt="Screenshot 2024-05-17 at 14 49 08" src="https://github.com/paritytech/polkadot-sdk/assets/2580396/ff1d5d5e-ce1a-4211-9ee4-f901c1a01b99">

pgherveou commented:
> Yes you are right. The storage accesses look correct. They were too low before. One last thing. There should be only one access to `Account` for `seal_instantiate` to bring it in line with `seal_call`.
> <img alt="Screenshot 2024-05-17 at 14 49 08" width="1369" src="https://private-user-images.githubusercontent.com/2580396/331598809-ff1d5d5e-ce1a-4211-9ee4-f901c1a01b99.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NTYxNTQsIm5iZiI6MTcxNTk1NTg1NCwicGF0aCI6Ii8yNTgwMzk2LzMzMTU5ODgwOS1mZjFkNWQ1ZS1jZTFhLTQyMTEtOWVlNC1mOTAxYzFhMDFiOTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTdUMTQyNDE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGM0MjBmOGMxMTdlNzE3YTAxN2I1Mzc4OGMwNDgzZDY2MGNiZGUwNDViNTJkODVhZGI4YzI3MmYwYWY5ODY0ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FCe-5EjWGxe7GhZyeuHE2hc-KKai99-IEoUmj-uwSPQ">

Yes reason for that is that the bench_call setup is using the contract caller as origin, I am changing it to the root_contract for delegate/call/instantiate, so that the origin account get whitelisted

pgherveou commented:
bot bench substrate-pallet --pallet=pallet_contracts

athei reviewed (commented):
Something still doesn't add up with the `seal_instantiate` benchmark. The base weight is super low (lower than `seal_call`)  and then a lot is added in case of transfer:
<img width="1321" alt="Screenshot 2024-05-18 at 07 31 52" src="https://github.com/paritytech/polkadot-sdk/assets/2580396/8cf53f53-7e90-4b02-a22c-71998ef24ac6">

pgherveou commented:
> Something still doesn't add up with the `seal_instantiate` benchmark. The base weight is super low (lower than `seal_call`) and then a lot is added in case of transfer: <img alt="Screenshot 2024-05-18 at 07 31 52" width="1321" src="https://private-user-images.githubusercontent.com/2580396/331768399-8cf53f53-7e90-4b02-a22c-71998ef24ac6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYxNDY1MDYsIm5iZiI6MTcxNjE0NjIwNiwicGF0aCI6Ii8yNTgwMzk2LzMzMTc2ODM5OS04Y2Y1M2Y1My03ZTkwLTRiMDItYTIyYy03MTk5OGVmMjRhYzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTlUMTkxNjQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTBlYjk1NDM0MzM2MzFlNzk5NmVhYzU2ZDg4MGUzMGE3NGIyZDYxN2ViNmUwZTA1NThkYWM2MWQ2ZmQzNjhmMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.klNVvGLCRIWglENCLGH-Dvnlp4Pee1P83H9gdUEoNsY">

mm interesting, will look into this tomorrow.
Relaunching one more time to check if this is consistent

bot bench substrate-pallet --pallet=pallet_contracts

pgherveou commented:
bot bench substrate-pallet --pallet=pallet_contracts

pgherveou commented:
> Something still doesn't add up with the `seal_instantiate` benchmark. The base weight is super low (lower than `seal_call`) and then a lot is added in case of transfer: <img alt="Screenshot 2024-05-18 at 07 31 52" width="1321" src="https://private-user-images.githubusercontent.com/2580396/331768399-8cf53f53-7e90-4b02-a22c-71998ef24ac6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyMTQxODMsIm5iZiI6MTcxNjIxMzg4MywicGF0aCI6Ii8yNTgwMzk2LzMzMTc2ODM5OS04Y2Y1M2Y1My03ZTkwLTRiMDItYTIyYy03MTk5OGVmMjRhYzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjBUMTQwNDQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTRmZjVkM2RhNzIxYjM5Njk3MjEwOTU1OWMwMTQ1ZjFhZWZhZTJmNmM5ZDdjZTY1MmYzY2UwYmYyOGRlM2RmNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.6YX0UvsAK_fBkWjmnHTofQlhtoIKxM3WLkYYkzcjy6M">

@athei Actually I think it does adds up
We are just forgetting the RocksDbWeight (which confusingly are actually adding ref_time and not pov)
on `seal_call` they are attached to `t` since  we only touch accounts if we do a transfer. 
On `instantiate` it happens all the time and does not depend on `t`

so if you count these 3 lines on `seal_call` 

```rust
.saturating_add(Weight::from_parts(44_066_869, 0).saturating_mul(t.into()))
.saturating_add(RocksDbWeight::get().reads((1_u64).saturating_mul(t.into())))
.saturating_add(RocksDbWeight::get().writes((1_u64).saturating_mul(t.into())))

// 44_000_000 + 25_000_000 + 100_000_000 ~=  169_000_000
```

You roughly get what we have in `seal_instantiate`
```rust
.saturating_add(Weight::from_parts(159_867_027, 0).saturating_mul(t.into()))
```

athei reviewed (approved):
You are right. I wasn't accounting for the reads and writes which are added unconditionally for `seal_instantiate`.

Robbepop reviewed (approved):
This change is very reasonable and I am glad that it has been done. The new resulting numbers look reasonable and should be more precise than before. Thanks @pgherveou !

xermicus reviewed (approved):
LGTM!
